### ISSUE TITLE
Write PHP module which makes use of RNP functionality.

### ISSUE THREAD

**Issue by ni4** [rnpgp/rnp#1739]
## Description

It appears that there are not much of OpenPGP implementations, available for the PHP language. Given that Drupal and Wordpress are quite popular as a site backend there could be some demand for PHP PGP module. Hopefully somebody from the PHP world would give us sime insights on this so we may prioritize this issue.

**Comment by antonsviridenko** [comment]
I can try to write such module, I had some experience with PHP module API.
However, PHP code can always call external command line tool, like `gpg` or `rnp`.

**Comment by ronaldtse** [comment]
@antonsviridenko that's great! Can you help do this?

**Comment by ni4** [comment]
> However, PHP code can always call external command line tool, like `gpg` or `rnp`.

This would not be a good approach for high-loaded servers, and may lead to some vulnerabilities, like this one: https://ubuntu.com/security/CVE-2018-12020

**Comment by antonsviridenko** [comment]
ok, how it should look like?
All API functions from <rnp/rnp.h> exported as close as possible to original C version?

**Comment by ni4** [comment]
> ok, how it should look like?
> All API functions from <rnp/rnp.h> exported as close as possible to original C version?

It should be something similar to `ruby-rnp`/`py-rnp` : first part is a `raw` lib which just imports C functions, and second is native PHP wrapper with classes, native types and so on. I doubt that PHP users will strictly follow FFI API rules for object and buffer destroying, use raw C strings and so on :)

**Comment by antonsviridenko** [comment]
Which PHP version should be targeted?
https://www.php.net/supported-versions.php
PHP 7.4 probably is going to be EOL-ed when this module becomes ready and reaches repositories.

**Comment by ronaldtse** [comment]
@antonsviridenko perhaps let's leave that decision to you since you are definitely more familiar with PHP than I am 😄

**Comment by antonsviridenko** [comment]
Actually 3rd party module for interaction with gnupg does exist.
https://pecl.php.net/package/gnupg
https://github.com/php-gnupg/php-gnupg

**Comment by ni4** [comment]
@antonsviridenko Yeah, but it uses gpgme which invokes gnupg executable. Also it would restrict user to the LGPL license.

**Comment by antonsviridenko** [comment]
How language bindings (py-rnp, ruby-rnp, swift-rnp) are synchronized with RNP API updates?

**Comment by ni4** [comment]
> How language bindings (py-rnp, ruby-rnp, swift-rnp) are synchronized with RNP API updates?

Manually :) There is no any formal process yet. Probably we should make one, or add some CI, so bindings would track newly added FFI functions.

**Comment by antonsviridenko** [comment]
Started to work on PHP module, 3 API functions are implemented so far :)
https://github.com/rnpgp/php-rnp

**Comment by ronaldtse** [comment]
Thank you @antonsviridenko !

**Comment by antonsviridenko** [comment]
Any ideas about password provider implementations for PHP? :)

If PHP script runs as a command line, it can use stdin password provider, like default in CLI,
but most frequent use is in context of web server process.

**Comment by ni4** [comment]
@antonsviridenko In web server context I see only two ways - via some callback (is it possible for PHP?), or via string constant. The second option however would need to know in advance to which key(s) or/and passwords data is encrypted. Probably we should extend our API with some `rnp_get_recipients()` call to be able to analyze input data for recipients.

**Comment by antonsviridenko** [comment]
Yeah, it appears callbacks are available in PHP.

https://www.php.net/manual/en/language.types.callable.php

Btw, how it is done in other language bindings?

**Comment by ni4** [comment]
@antonsviridenko ruby-rnp uses callback or predefined password approach, please see this example:
https://github.com/rnpgp/ruby-rnp/blob/main/examples/password_provider.rb

**Comment by antonsviridenko** [comment]
Let's talk about the license of this module. If we want it to be included into the PECL, "The PHP License, version 3.01" is preferred.

https://pecl.php.net/account-request.php

**Comment by ronaldtse** [comment]
@antonsviridenko it looks like they support the BSD license too? Let’s use the same BSD license for simplicity in managing our project. Thanks!

**Comment by antonsviridenko** [comment]
@ronaldtse ok, added LICENSE file with Ribose license text https://github.com/rnpgp/php-rnp/blob/main/LICENSE

**Comment by antonsviridenko** [comment]
So, minimal amount of required APIs for key generation/sign/verify/encrypt/decrypt operations is implemented so far. PHP API looks like this now - https://github.com/rnpgp/php-rnp/blob/main/rnp.stub.php. Time to write some PHP documentation :)

**Comment by antonsviridenko** [comment]
Keys are referenced using fingerprint strings or arrays containing such strings (`$keys_fp, $recipient_keys_fp, return value of rnp_op_generate_key()`)

**Comment by ronaldtse** [comment]
Thank you @antonsviridenko !

**Comment by antonsviridenko** [comment]
@ronaldtse Did you receive any response from PECL-Dev? Probably you need to subscribe to this mailing list first, your email is not visible in mailing list archives, probably it did not reach them 

https://marc.info/?l=pecl-dev&r=1&b=202208&w=2

**Comment by ronaldtse** [comment]
@antonsviridenko you're absolutely right -- I've just re-done it and the message is now on the list. Thanks!

**Comment by antonsviridenko** [comment]
@ronaldtse any news about PECL account?

**Comment by antonsviridenko** [comment]
What version number should we assign to the first release of this module? PECL requires versioned releases as I see.

**Comment by antonsviridenko** [comment]
Let it be version 0.1

**Comment by antonsviridenko** [comment]
@ronaldtse did you register the PECL account?

As I see, there are two existing approaches,
one is that each developer has individual PECL account under his name
and multiple PECL accounts can be listed in package as maintainers of this package,
or there is a team account named like "Protobuf Team",
so we can have one named like "RNP Team".

PHP module is almost ready to release, finishing documentation,
we can release it on PECL soon if everything is ok.

**Comment by ronaldtse** [comment]
@antonsviridenko thank you for the reminder! We haven't heard back since Sept from the PECL-DEV mailing list, so I just submitted an account creation request, and emailed the list (and Christoph Baker) again to request for approval. Will let you know!

**Comment by ronaldtse** [comment]
(Agree with version 0.1) Once we have the team account perhaps we can also have individual PECL accounts as maintainer.

**Comment by antonsviridenko** [comment]
@ronaldtse great, I see they've approved your PECL account, can you tell me PECL username you registered? It should be included into the https://github.com/rnpgp/php-rnp/blob/main/package.xml file for successfull upload.

**Comment by ronaldtse** [comment]
@antonsviridenko the username is `rnp`. Is there anything else we need?

**Comment by antonsviridenko** [comment]
@ronaldtse ok, I've updated https://github.com/rnpgp/php-rnp/blob/main/package.xml with actual username, release date and full file list, now you can use your PECL account & try to upload the first release using this file. Let me know if something is not ok.

**Comment by antonsviridenko** [comment]
After the package is listed in PECL, I can create pull requests to PHP documentation repositories, link to existing PECL package is required for approval.
